### PR TITLE
mgr/dashboard: fix POST alerts request

### PR DESCRIPTION
**Pull request by avanthakkar**
Fixes: https://tracker.ceph.com/issues/51360
Signed-off-by: Avan Thakkar <athakkar@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by nizamial09 — APPROVED** [review]

**Review by epuertat — APPROVED** [review]

**Review by epuertat — COMMENTED** [review]
Could you please add an API test to ensure we catch this if it happens again?

**Comment by avanthakkar** [comment]
Currently the POST alerts request with empty body sends the notification too (as checked in the unit-test). So should we allow this or only allow to POST if `notifications` dict is not `None` ? @alfonsomthd @epuertat

**Comment by alfonsomthd** [comment]
> Currently the POST alerts request with empty body sends the notification too (as checked in the unit-test). So should we allow this or only allow to POST if `notifications` dict is not `None` ? @alfonsomthd @epuertat

@avanthakkar IMHO we should not allow empty notifications, but throw a 400 error.

File: src/pybind/mgr/dashboard/tests/test_prometheus.py [hunk 1]
```diff
@@ -79,6 +79,11 @@ def test_post_on_receiver(self):
         self.assertEqual(notification['name'], 'foo')
         self.assertTrue(len(notification['notified']) > 20)
 
+        # test with empty body
+        self._post('/api/prometheus_receiver', {})
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/tests/test_prometheus.py:83]: Are you sure this is an empty request? Actually it has a body with an empty object, which is not the same as an empty body... To trigger this issue this should send a request without any data in the body at all.

Comment (epuertat) [reply src/pybind/mgr/dashboard/tests/test_prometheus.py:83]: Actually the fact that we are now triggering an exception and this test passes without catching that exception means that the test is not covering the code path you fixed.

**Review by alfonsomthd — DISMISSED** [review]

**Comment by avanthakkar** [comment]
I think PR should be closed as `allow_empty_body` (the name is the confusion actually) is used if we don't want to throw 415 error is case if the request is made with content type other than wht is specified in `Content-Type` header. But I think here that's not the case, it completely works if correct JSON payload is passed in the request body along with specifying the expected type in request `Content-Type` header. And regarding the empty notifications, I don't think prometheus does send any, it's just I assumed in case if any user directly consumes the endpoint then it should not be allowed to `POST` with 400 error code.

**Comment by epuertat** [comment]
As clarified with @avanthakkar, this is allowing non-standard usage of HTTP POST without proper Content-type or empty body. Closing then, as it's working with Prometheus Alertmanager.

**Pull request closed without merging**